### ISSUE TITLE
zeKernelGetSourceAttributes signature is inconsistent

### ISSUE THREAD

**Issue by Kerilk** [oneapi-src/level-zero-spec#92]
The specification for `zeKernelGetSourceAttributes` has been updated, see:
https://github.com/wdamon-intel/level-zero-spec/blob/434bb895f4d1838aaf3d4780f058617338c3ad4a/scripts/core/module.yml#L547-L566

But the signature hasn't been changed, it is still:
```C
ze_result_t zeKernelGetSourceAttributes(ze_kernel_handle_t hKernel, uint32_t* pSize, char** pString);
```
So if I am understanding the specification correctly, the user should write:

```C
uint32_t sourceAttributeSize = 0;
char * sourceAttributeBuffer = NULL;
ZE_CHECK(zeKernelGetSourceAttributes(kernel, &sourceAttributeSize, NULL));
sourceAttributeBuffer = (char *)malloc(sourceAttributeSize);
CHECK_ALLOC(sourceAttributeBuffer);
ZE_CHECK(zeKernelGetSourceAttributes(kernel, &sourceAttributeSize, sourceAttributeBuffer));
```

But the type of the `pString` parameter is `char **` and not `char *` so the above code will most probably fail to build.
`char **` made sense when the function was supposed to return an internal string (though the size parameter in this case did not), but doesn't now that the function is supposed to fill a user provided array. I don't know what's the best solution is here (deprecation, ABI change, ...) but if the semantic of the function can change this much, maybe changing the pointer type is not so much of an issue.

**Comment by wdamon-intel** [comment]
Yes, this is an unfortunate typo in the original signature. Your understanding is correct... using the example above, the only difference would be:

`ZE_CHECK(zeKernelGetSourceAttributes(kernel, &sourceAttributeSize, &sourceAttributeBuffer));`

(Note, this is what SYCL is doing. Please see here https://github.com/intel/llvm/blob/a0d0942d2b42d937b9ec41e7526d93c608a2beaf/sycl/plugins/level_zero/pi_level_zero.cpp#L4201)

It was never the intention for the implementation to provide an internal string. While changing the semantics of the function only requires change in documentation, the signature cannot be changed unless/until there is a major version release, in order to avoid breaking the ABI, as you mentioned. Right now, we are waiting for community feedback as to whether a new API is needed with a corrected signature or if keeping current one, with the caveat that the pointer to a pointer needs to be passed in, is an acceptable inconsistency.

**Comment by Kerilk** [comment]
Thanks for the clarification. As long as the ABI was kept stable and this is just a documentation problem, I can take appropriate action in THAPI tu update our behavior is required.

> Right now, we are waiting for community feedback as to whether a new API is needed with a corrected signature or if keeping current one, with the caveat that the pointer to a pointer needs to be passed in, is an acceptable inconsistency.

I would keep it as is until the next major version, but I would indicate the issue in the documentation, and that this methods does not follow the API best practice.

**Comment by rwmcguir** [comment]
Documentation has been updated calling this out...
